cmd/minishift/cmd/hostfolder/add.go

```diff
@@ -20,45 +20,231 @@ import (
+	"errors"
+	miniConfig "github.com/minishift/minishift/pkg/minishift/config"
+	hf "github.com/minishift/minishift/pkg/minishift/hostfolder"
+	"github.com/minishift/minishift/pkg/minishift/hostfolder/config"
+	"github.com/minishift/minishift/pkg/util"
 	"github.com/minishift/minishift/pkg/util/os/atexit"
+	"github.com/minishift/minishift/pkg/util/strings"
 	"github.com/spf13/cobra"
 )
 
+const (
+	usage                = "Usage: minishift hostfolder add --type TYPE --source SOURCE --target TARGET HOST_FOLDER_NAME"
+	noSource             = "you need to specify the source of the host folder"
+	noTarget             = "you need to specify the target of the host folder"
+	noUserName           = "you need to specify a username"
+	noPassword           = "you need to specify a password"
+	noDomain             = "you need to specify the Windows domain"
+	unknownType          = "'%s' is an unknown host folder type"
+	nonSupportedTtyError = "not a tty supported terminal"
+	shareTypeFlag        = "type"
+	sourceFlag           = "source"
+	targetFlag           = "target"
+	optionsFlag          = "options"
+	interactiveFlag      = "interactive"
+	instanceOnlyFlag     = "instance-only"
+	usersShareFlag       = "users-share"
+)
+
 var (
 	instanceOnly bool
 	usersShare   bool
+	interactive  bool
+	shareType    string
+	source       string
+	target       string
+	options      string
 )
 
-var hostfolderAddCmd = &cobra.Command{
-	Use:   "add HOSTFOLDER_NAME",
-	Short: "Adds a host folder definition.",
-	Long:  `Adds a host folder definition. The defined host folder can be mounted to a running OpenShift cluster.`,
-	Run: func(cmd *cobra.Command, args []string) {
+var addCmd = &cobra.Command{
+	Use:   "add HOST_FOLDER_NAME",
+	Short: "Adds a host folder config.",
+	Long:  `Adds a host folder config. The defined host folder can be mounted into the Minishift VM file system.`,
+	Run:   addHostFolder,
+}
 
-		var err error = nil
-		if usersShare && runtime.GOOS == "windows" {
-			// Windows-only (CIFS), all instances
-			err = hostfolderActions.SetupUsers(true)
-		} else {
-			if len(args) < 1 {
-				atexit.ExitWithMessage(1, "Usage: minishift hostfolder add HOSTFOLDER_NAME")
-			}
-			err = hostfolderActions.Add(args[0], !instanceOnly)
+func init() {
+	HostFolderCmd.AddCommand(addCmd)
+	addCmd.Flags().StringVarP(&shareType, shareTypeFlag, "t", "cifs", "The host folder type. Allowed types are [cifs|sshfs]. SSHFS is experimental.")
+	addCmd.Flags().StringVar(&source, sourceFlag, "", "The source of the host folder.")
+	addCmd.Flags().StringVar(&target, targetFlag, "", "The target (mount point) of the host folder.")
+	addCmd.Flags().StringVar(&options, optionsFlag, "", "Host folder type specific options.")
+	addCmd.Flags().BoolVar(&instanceOnly, instanceOnlyFlag, false, "Defines the host folder only for the current Minishift instance.")
+	addCmd.Flags().BoolVarP(&interactive, interactiveFlag, "i", false, "Allows to interactively provide the required parameters.")
```

Comment:
This changes the current behaviour and is NOT self-explanatory to those who already use it

Reply:
True, this changes the current behaviour. However, I think it was the 'hostfolder add' which was the odd one. Really the default mode for all commands should be non-interactive. 

I don't think this should be a big issue for anyone who is using CIFS already. It is also more connected to #959 than #317.

Eventually, I would not only make the default non-interactive but also switch the default type from CIFS to SSHFS. That's after we gained some experience with it.

@minishift/minishift-dev ^^ any more thoughts on what the default should be?

Reply:
If #959 would solve the interactive flow, than maybe it should not have been addressed here. Just my 2c as it does change bahaviour.

Reply:
Eventually we will suggest SSHFS, and later move to SSHFS as default, But for now, we have been experiencing different results on platforms, from unexplained errors to super-pleasant experience.

Reply:
> If #959 would solve the interactive flow, than maybe it should not have been addressed here. 

It was always the plan to address the interactive vs non-ineractive flow as part of the issue. It was a natural fall-out of the refactoring. Both modes are there. One just needs to set the default for the `interactive` flag to true and you have the old behaviour. The question is what makes most sense. IMO non-interactive. However, if making the default interactive gets us closer to merging this, I can do it. I was basically hoping for some more opinion on this one (@minishift/minishift-dev ). ATM, it seems to be 1:1. Anyone wants to weigh in to push the scale one way or the other?